types/row_proof.go

```diff
@@ -34,6 +34,11 @@ func (rp RowProof) Validate(root []byte) error {
 	if len(rp.Proofs) != len(rp.RowRoots) {
 		return fmt.Errorf("the number of proofs %d must equal the number of row roots %d", len(rp.Proofs), len(rp.RowRoots))
 	}
+	if len(rp.Proofs) != 0 &&
```

Comment:
Not only should the `Index` of the first and last proof match with the `StartRow` and `EndRow`, but also, if we aim to treat `Proofs` as an ordered list, then the `Index` of all the other proofs in `Proofs` should be sequential. 

BTW, I think this PR is somewhat a breaking change, since an unordered `Proofs` could have potentially been considered valid, but with this change, it no longer is (I'd suggest developing a test to see if this constitutes a breaking change).

Reply:
currently, all the constructed proofs are sequential. Do you want to add a validation for it too?

Reply:
My point is to be specific about what we want to enforce with this check: is the list sequential, or something else? If we only check the index of the first and last proof, nothing specific can be guaranteed about the order of proofs in `Proofs` since the middle ones can be in any arbitrary order (as they are not checked). Therefore, if we want to enforce the proofs to be ordered, we need to add a check for the middle ones too; otherwise, checking only the first and last does not guarantee anything. Hope that clarifies my comment.

Reply:
[aac1164](https://github.com/celestiaorg/celestia-core/pull/1476/commits/aac11640e30278fb60d1fe8fab6efcaa3a1d3f33)